Gallery block / media-placeholder - Preserve changes made while upload in progress.

Addison-Stavlo opened the pull request:
## Description
<!-- Please describe what you have changed or added -->
This PR aims to fix 3 bugs related to the Gallery block component.

### Before - The bugs that exist.

#### 1 - When Images are removed while another is uploading, they are re-added once upload is complete.
![master-caseC](https://user-images.githubusercontent.com/28742426/70823213-5886d300-1dad-11ea-9704-72779f54dd4a.gif)

#### 2 - When two images are uploaded separately, if the first finishes after the second then the second one is removed once the first finishes its upload.
![master-caseA](https://user-images.githubusercontent.com/28742426/70823317-94ba3380-1dad-11ea-9c0c-7457e6480f19.gif)

#### 3 - When two images are uploaded separately, if the first finishes before the second then the loader for the second disappears.  Once the second finishes, "empty" loading image replaces the first.
![master-caseB](https://user-images.githubusercontent.com/28742426/70823400-ccc17680-1dad-11ea-8eaa-44cbf55846b2.gif)

### How do we go about fixing the above bugs?

Currently, the `media-placeholder` component destructures the `value` from props and adds this into the `setMedia` function as `currentValue`.  This is never updated during the lifetime of the function which is why the above issues exist.  In case 1, since these images are initially added to the function in `currentValue`, they remain in `currentValue` even after they are deleted from the screen.  Once the `setMedia` function runs when the uploading image completes, it returns `currentValue` with those unwanted deleted images.

To fix this, I have added another reference to `this.props.value` (named `currentMedia`) but declared inside the `setMedia` function itself.  This allows us to compare `currentValue` (the 'snapshot' of the gallery's images at the beginning of the upload, now renamed `mediaToReturn` ) with the values that are in the gallery at the time the function is run.

This now allows us to do 2 important checks when the `setMedia` function runs:

1. Have items been removed from the gallery since the upload started?
2. Have completed items been added to the gallery since the upload started?

With these checks we are then able to add/remove items from `mediaToReturn` as necessary so that once the upload completes, we are able to return a state representative of changes that have happened since it started.

##### Note: Why update `mediaToReturn` and not just use the current values in `this.props.value` altogether?
You will find that `this.props.value` contains temporary items returned from the `setMedia` function, such as the loading placeholders.  We don't want to return these placeholders with their corresponding completed uploads, so they would need to be filtered out in that case.  Since currently there is not a good way to tell which placeholders correspond to which completed uploads, it is more simple to add items to `mediaToReturn` than to remove the correctly corresponding ones from the current value of the gallery.

### After the described changes.

#### 1 - When Images are removed while another is uploading, they are not re-added once the upload completes.
![branch-caseC](https://user-images.githubusercontent.com/28742426/70823439-e5319100-1dad-11ea-997b-6fc2422af3e7.gif)

#### 2 - When two images uploaded separately, if the first finishes after the second they will both remain in the gallery.
![branch-caseA](https://user-images.githubusercontent.com/28742426/70823533-1ad67a00-1dae-11ea-86e0-be61fbb57082.gif)

#### 3 - When two images are uploaded separately, if the first finishes before the second the loader for the second still disappears.  However, once it finishes its upload they both are present in the gallery with no empty loading placeholder.
![branch-caseB](https://user-images.githubusercontent.com/28742426/70823608-48232800-1dae-11ea-9452-b1a2ceaff029.gif)

### Some comments on the code itself

1.  There is an adjactent `*.native.js` file to the `media-placeholder` index file.  This has yet to be updated as I am not currently set up to test functionality on native and wanted to get the discussion on this moving before then.  None of the changes made should break any functionality on native, but since the logic in the native file has not been updated there should be no change to how the gallery block works on native until this is similarly updated.  It may also be worth noting that cases 2 & 3 may be much less commonly experienced through the mobile workflow.

2. For comparisons we are using `id`s (and in some cases `url`s) of the media.  My first thought was to use lodash `isEqual` to compare the objects, but noticed there could be some inconsistencies on which properties were present on the objects throughout the lifetime of the function ( `fullUrl: undefined` for example may be present in the values on the first pass of the function, but absent from the copies returned with the loading placeholder causing `isEqual` to always return `false` on a second pass. ).

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Tested on local gutenberg docker environment.
Run this PR and test initiating different uploads simultaneously, removing images from the gallery during an upload, etc.  Recreate the above scenarios listed and test with group uploads as well.  

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

Addison-Stavlo committed:
added optional prop to pass by reference instead of value

Addison-Stavlo committed:
currentValue updated on adding/removing

Addison-Stavlo committed:
comment update

Addison-Stavlo committed:
fixed multiple upload bug

Addison-Stavlo committed:
removed stuff for trying to keep loaders that werent part of the upload group

Addison-Stavlo committed:
updates comments and naming

Addison-Stavlo committed:
updated galleryValue to fallback to empty array

noahtallen commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -118,12 +118,31 @@ export class MediaPlaceholder extends Component {
+						const galleryValue = this.props.value || [];
+						// If currentValue contains things not in galleryValue, remove them.
+						currentValue = currentValue.filter( ( item ) => {
+							return galleryValue.some( ( propItem ) => propItem.id === item.id && propItem.url === item.url );
```
One thing I found confusing is the naming `currentValue` vs `galleryValue`. Would it make more sense to say something like `incomingValue`?

noahtallen replied:
or newValues or something like that

Addison-Stavlo replied:
I agree!  We should probably update this, I had been doing some debating about how to properly rename these but ended up sending myself in circles.

Essentially, we should rename `galleryValue` to be `currentValue`, as it does represent the current value that is in place in the gallery block.  Then `currentValue` could be named something like `initialValue`, since it is initially defined at the start of the upload.  But then `initialValue` seems to be a bad name for it since we are updating its contents when the function runs to no longer be the initial value... LOL..

What about renaming `galleryValue` to `currentValue`, and rename `currentValue` as `returnValue` ?  Even that name has some issues since `newMedia` is `.concat()`d onto that at the end, so it isn't really the full return value...  I am open for suggestions on what to rename `currentValue` to if you have any ideas?

noahtallen replied:
Hm -- I feel like I still get confused by that. It may be worth not saying `Value`, but saying `Images` or `Uploads`. Just something to nail down exactly what we're talking about. Something like `currentUploads` or `currentImages`. I feel like `returnValues` is confusing, but maybe more along that line of thinking like `finalUploads`.

Is the difference here something like `pendingUploads` (for `galleryValue`) and then `completedUploads`? that would be a lot clearer to me, but I'm not sure if that's actually what the code is doing

Addison-Stavlo replied:
> that would be a lot clearer to me, but I'm not sure if that's actually what the code is doing

Thats not what its doing unfortunately, but I think something like `currentMedia` and `mediaToReturn` would be a bit better than what is currently there?  `currentMedia` would be as we see the state of `this.props.value` when the function runs, and `mediaToReturn` would initially be assigned as `value` and updated along the lifecycle of the function.

noahtallen commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -118,12 +118,31 @@ export class MediaPlaceholder extends Component {
+							return galleryValue.some( ( propItem ) => propItem.id === item.id && propItem.url === item.url );
+						} );
+						// If galleryValue has completed items not in currentValue, add them.
+						galleryValue.forEach( ( propItem ) => {
```
Similarly to above,  what if instead of `propItem`, we said `newItem`?

Addison-Stavlo replied:
I updated it to be `currentItem` to fit with `currentMedia` and `mediaToReturn`, although `currentItem` still seems kind of odd it does at least make sense being an item from `currentMedia`.

noahtallen reviewed (approved):
Thanks for the super detailed PR description! There were several times I had questions about the code that were answered in the description. Very nice :)

> we are using ids of the media

This makes sense to me. Is there a reason why we use both ID and URL, or could we just use one of those?

>  there should be no change to how the gallery block works on native until this is similarly updated

I'm also not sure how to develop the changes for the react native side of things. My thought would be to merge this before looking into changes to native code.

> This may not be necessary but has been done to ensure backwards compatibility with all other components that use this media-placeholder.

Ahh this makes a lot of sense. Do you know if we experience similar bugs on other components that would be fixed once remove this check?

I did notice one more bug which happens when I do the following:
1. Start with an empty gallery
2. Add several images at the same time
3. When the first image finishes uploading but before the others have finished uploading, click into it and remove it.
4. Once the other images upload, you'll see the removed image re-appear in the gallery.

Overall though, great bug fix. I think it's ready to merge unless you want to fix any of the nitpicks! :)

youknowriad commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -118,12 +118,31 @@ export class MediaPlaceholder extends Component {
 			onError,
 			onSelect,
 			value = [],
+			dynamicAlteration,
```
Can we explain a bit more what this PR does? Is it needed? should we just always use that behavior?

Addison-Stavlo replied:
It is currently only in place as a safety measure to ensure this code is only running for the Gallery block.  I did not want to immediately throw it into use in all cases as I may be missing a use of it somewhere and did not want to accidentally break anything that I haven't been able to explicitly test.  It may not be needed in the end...

youknowriad replied:
I'd rather personally be more conservative about introducing new APIs. Each time we introduce something new (like this prop) we're committing to ensure backward compatibility for this prop. (if it's still in the plugin, it's two releases but once it can be forever).

So if we really think, this should be the default behavior, we should just avoid the prop.

Addison-Stavlo replied:
Great points!

I have removed the `dynamicAlteration` prop.  After searching through the codebase once more for all uses of the `MediaPlaceholder` component, it seems that the `Gallery` component is the only use of the `MediaPlaceholder` where the `multiple` property used (If this is not the case I would love to hear where else it is used, I have not been able to find anything).  So even without my unnecessary prop, it seems that the `Gallery` block is still the only block effected by this change.

youknowriad replied:
Thanks

Addison-Stavlo commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -118,12 +118,31 @@ export class MediaPlaceholder extends Component {
+				let currentValue = value;
 				setMedia = ( newMedia ) => {
+					if ( dynamicAlteration ) {
+						const galleryValue = this.props.value || [];
```
Unfortunately that will not work.  If we define it to be the same as `value`, then `galleryValue` will be what `value` was at the time the function was defined and passed on for use by `mediaUpload`, essentially making it the same a `currentValue` which caused these bugs in the first place.

What is important about this is that we are assigning this from `this.props` inside the function itself.  Meaning that whenever the function runs it is defining `galleryValue` based off what is currently happening in the gallery block (as opposed to what we had destructured from props at the time when we defined the function).

noahtallen replied:
ahh I see what you mean. That makes sense, thanks!

Addison-Stavlo commented:
> Is there a reason why we use both ID and URL, or could we just use one of those?

@noahtallen Yeah, now that you mention it we may not need the `Id`s in that first check.  I added the `url`s as a second thought later.  Since placeholder images don't come back from the Gallerys props with `id`s on them, I needed to add something else so they were not checking `undefined === undefined`.

> Do you know if we experience similar bugs on other components that would be fixed once remove this check?

I am not completely sure, I will do some more searching on this.  I think the best thing will be to adapt this to not require the `dynamicAlteration` prop and use this logic by default in the `multiple` use case.  I just need to make sure we test ALL of the other components that would be effected by it and adapt it accordingly.

> I did notice one more bug which happens when I do the following:
Start with an empty gallery
Add several images at the same time
When the first image finishes uploading but before the others have finished uploading, click into it and remove it.
Once the other images upload, you'll see the removed image re-appear in the gallery.

Yeah, there are a couple other odd bugs like this that exist.  While this does not fix them, it does not create them either.  I think at that point since it was triggered by a group upload, it is less likely that the user will want to remove it during the group upload.  What is more likely is what we are fixing where a user may select a group upload and remove images (that were already on the gallery) while that upload is in progress.  I think to fix the last of these bugs we may need to reorganize both how `mediaUpload` passes `newMedia` into the `setMedia` function as well as how the placeholders are set in the Gallery block, which might be a bit larger in scope.

Addison-Stavlo committed:
renamed values to be more semantic

youknowriad reviewed (commented):
@jorgefilipecosta do you have thoughts here?

Addison-Stavlo committed:
removed dynamicAlteration prop

jorgefilipecosta commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -122,9 +122,25 @@ export class MediaPlaceholder extends Component {
 			if ( addToGallery ) {
-				const currentValue = value;
+				let mediaToReturn = value;
 				setMedia = ( newMedia ) => {
```
Hi @Addison-Stavlo,
Thank you for debugging these issues and proposing a solution to this problem 👍
This solution seems to work well.
I wonder if it would be possible to simplify the logic a bit by always containing the newMedia with this.props.value with all blob URLs filtered (removed) using isBlobUrl function in the filter from @wordpress/blob? Do you think the logic would be equivalent?

Addison-Stavlo replied:
Hi @jorgefilipecosta ,
Thanks for taking the time to give some feedback on this.  

> I wonder if it would be possible to simplify the logic a bit by always containing the newMedia with this.props.value with all blob URLs filtered (removed) using isBlobUrl function

It seems that this would remove blobs from other upload groups, add duplicates, and possibly introduce some other unwanted side effects that we would need to figure out how to get around.  However, we can use this idea of always concating `newMedia` onto a filtered `this.props.value` but instead filter out any media that corresponds to `newMedia` given by the last pass of the function.  

This should both clean up the logic a little bit as well as fix the remaining bug noted in example gif 3.

Addison-Stavlo committed:
refactored to compare last pass of newMedia

Addison-Stavlo commented:
I have updated the logic a little bit here.  It is now a little more simplified and robust.  This now also fixes the remaining issue I noted above in example 3.

> 3 - When two images are uploaded separately, if the first finishes before the second the loader for the second still disappears. However, once it finishes its upload they both are present in the gallery with no empty loading placeholder.

Now, the second loader will no longer disappear once the first upload completes.

![ex3-fixed](https://user-images.githubusercontent.com/28742426/71747976-2e21b600-2e3f-11ea-9a03-b901f7828299.gif)

Here we take `this.props.value` and filter out anything we had added with `newMedia` in the last run of the function.  We return this with the current value of `newMedia` added to it, and save these URLs for filtering again in the next iteration.

To do this I set a `lastMediaPassed` array just before the `setMedia` function declaration.  When `setMedia` runs, it takes the `currentMedia` from `this.props.value` and filters items out whose URLs correspond to those in `lastMediaPassed`, this is then returned along with `newMedia`.  `lastMediaPassed` is then reset in each run of the function and the URLs from `newMedia` are added to it to allow comparison the next time the function runs.

noahtallen commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -117,14 +117,22 @@ export class MediaPlaceholder extends Component {
+						return ! lastMediaPassed.some( ( temporaryMediaURL ) => temporaryMediaURL === item.url );
+					} );
+					// Reset lastMediaPassed and set it with urls from newMedia.
+					lastMediaPassed = [];
```
Could we just do this if `lastMediaPassed` is just being set to all the urls in newMedia? might be a bit clearer

```suggestion
					lastMediaPassed = newMedia.map( ( media ) => media.url );
```

Addison-Stavlo replied:
Definitely!  Updated to do this =)

noahtallen commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -117,14 +117,22 @@ export class MediaPlaceholder extends Component {
+					lastMediaPassed = [];
+					newMedia.forEach( ( media ) => lastMediaPassed.push( media.url ) );
+
+					onSelect( mediaToReturn.concat( newMedia ) );
```
Could we add a comment here to explain what the variable means -- like why we use `mediaToReturn` instead of `lastMediaPassed` or something? It just took me a while to understand what the different variables were doing again, so a comment might help make it clearer to new folks

Addison-Stavlo replied:
I added a few more comments and renamed `mediaToReturn` as well.  Hopefully this reads a bit easier now?

noahtallen replied:
Yeah, that definitely reads better, thanks! :D

noahtallen reviewed (commented):
nice changes! I think this is almost ready to go

Addison-Stavlo committed:
added comments, changed variable name

Addison-Stavlo committed:
simplified setting lastMediaPassed

noahtallen commented:
I think this might need a rebase to get the test passing again :)

Addison-Stavlo commented:
> I think this might need a rebase to get the test passing again :)

I did the rebase, but it seems the author e2e test `Buttons › can jump to the link editor using the keyboard shortcut` is still failing due to a snapshot mismatch.  This seems unrelated?

aduth commented:
The build has been more unstable than usual lately, so it's quite likely to be unrelated. I've restarted the build to try to clear it up.

Addison-Stavlo commented:
@jorgefilipecosta & @youknowriad - What are your thoughts on how this is looking now?  The logic is a bit more simplified and it has a better result in terms of behavior as well.  What else might we need to do to get this fix accepted?

jorgefilipecosta reviewed (commented):
Hi @Addison-Stavlo thank you for your changes the new logic is much simpler 👍 
Unfortunately, it seems there is small regression, when I create a gallery with images from the media gallery, and then I drag & drop (to upload) two images into the gallery one of the images is added two times into the gallery.

Addison-Stavlo committed:
compare via includes instead

Addison-Stavlo commented:
@jorgefilipecosta - Wow, im sorry I missed that.  Thanks for pointing that out.

It looks like `newMedia` in the `setMedia` function is passing the finalized image with url of `some-name.png`, but once it gets placed into the gallery its url is then `some-name-1024x724.png` (or whatever size it happens to be).  Because of this mismatch the `setMedia` function isn't filtering out completed images that its upload group is responsible for, resulting in the duplicate noted.

Instead of filtering based on `some-name-ratio.png` === `some-name.png` I have updated it to compare via `string.includes`, so something like `'some-name-ratio.png'.includes( 'some-name' )`.  This seems to fix it and I'm thinking `some-name` should be unique enough that this method shouldn't cause any issues.

jorgefilipecosta commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -116,14 +116,38 @@ export class MediaPlaceholder extends Component {
+					// Reset lastMediaPassed and set it with urls from newMedia.
+					lastMediaPassed = newMedia.map( ( media ) => {
+						// Add everything up to '.fileType' to compare via .includes.
+						const cutOffIndex = media.url.lastIndexOf( '.' );
```
If the user tries to upload two different images one named image-267x300.jpg and another named image.jpg, is there any chance a problem would occur?

Addison-Stavlo replied:
Great thought!

It looks like there is a problem when `some-name-X` is already uploaded, and then `some-name` is uploaded as a not-last member of a group upload.  But your idea mentioned below regarding comparing by ID if it exists seems to fix this case.

jorgefilipecosta commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -116,14 +116,38 @@ export class MediaPlaceholder extends Component {
+					// Their replacements are contained in newMedia.
+					const filteredMedia = currentMedia.filter( ( item ) => {
+						// Compare via .includes since gallery adds a ratio to the url after returned from this function.
+						return ! lastMediaPassed.some( ( temporaryMediaSlug ) => item.url.includes( temporaryMediaSlug ) );
```
This comparison relies on WordPress using a specific format for the filenames of different image sizes. Is there any possibility a plugin could change this format?
Would it be possible for lastMediaPassed be an object that contains id and URL and when the id is available we perform the comparison by id instead of URL to avoid the need to rely on WordPress images with the same id have a common URL prefix?

Addison-Stavlo replied:
While I am not sure about the plugins changing the format, I give a huge YES to adding the check by Id if it is available.  This should force the url prefix check to only happen on transient images that have not been fully uploaded.

After adding that Id check, the only edge-case I can think of is if `image-name-something` and `image-name` being uploaded simultaneously in different upload groups. Even then, this would be extremely rare and would only temporarily get rid of a transient placeholder that would still be re-added (or inevitably replaced by the fully uploaded image once complete).

jorgefilipecosta reviewed (commented):
Hi @Addison-Stavlo, it seems the bug I identified is fixed 👍 I left some questions the main one being if we can use the id to compare images instead of URL as it would be more reliable against future changes in the way WordPress generates URLs for the different image sizes.

Addison-Stavlo committed:
added id check as first priority in filter

Addison-Stavlo commented:
@jorgefilipecosta - I'd like to say a huge THANK YOU for all your help on this one.  

I added the Id check you requested which seems to make this quite a bit more solid.

Addison-Stavlo committed:
edited comments

noahtallen commented:
@jorgefilipecosta  @Addison-Stavlo bump :)

lancewillett commented:
Another vote to bump this one back into the review queue. Still seeing users on WordPress.com struggle with this as they edit existing images in a Gallery block.

jorgefilipecosta reviewed (commented):
Hi @Addison-Stavlo, I'm sorry for the delay the notifications on this PR got lost in the pings I had.

I rebased this PR and gave it a new set of tests after all the changes that happened e.g: g2.
It seems the PR is not working after the rebase.
I created a custom network limitation to test the PR in the dev tools with a 5-second latency.
I added a gallery, I dragged an image to it, I waited to seconds (image still uploading) and then dragged another image. Only the last image is in the gallery when both uploads finish.

Addison-Stavlo committed:
rebased

Addison-Stavlo reviewed (commented):
Hi @jorgefilipecosta , 

> I rebased this PR and gave it a new set of tests after all the changes that happened e.g: g2.
It seems the PR is not working after the rebase.
I created a custom network limitation to test the PR in the dev tools with a 5-second latency.
I added a gallery, I dragged an image to it, I waited to seconds (image still uploading) and then dragged another image. Only the last image is in the gallery when both uploads finish.

~I see what you mean!  That is interesting.  It seems this is only the case before an initial image has been uploaded.  Once the gallery has 1 image uploaded, this logic and functionality seems to work again.  I am wondering if this is just an edge case we forgot to test previously, and not a result of other Gutenberg changes.~

Never mind on that thought, it seems other things (such as uploading groups of images) have gotten messed up as well.  This may need to be revisited with a different approach.

Addison-Stavlo committed:
changed to soft equal for int/string ID comparison

Addison-Stavlo committed:
added unchanging reference for image array

Addison-Stavlo reviewed (commented):
@jorgefilipecosta 
Things should be working now.

For the Issue you noted:
- This may have been an edge case we missed testing earlier, as it was only happening if there was not 1 image already uploaded.  To fix this I needed to make some changes in `packages/block-library/src/gallery/edit.js` to ensure we have an unchanging reference to the image array.  For whatever reason, when we go from 0 images to some images the array passed down in `props.attributes.images` is redefined, so the first uploading image was holding onto an outdated reference for the `currentMedia` of the gallery.

Besides this, I found a new issue when uploading groups of images.  For this I changed the id comparison to a soft `==` instead of `===`.  It seems recent changes have this circling back as a string where it has once been an integer (or vice versa), causing the filter to let duplicates through.

Addison-Stavlo commented:
Bumping this for when there is some time to revisit - @jorgefilipecosta

jorgefilipecosta commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -312,6 +314,14 @@ class GalleryEdit extends Component {
 	}
 
+	/**
+	 * Sets the images in our array without redifining reference.
```
Why we can not redefine the reference? It seems like we could remove this.images and simply pass this.props.attributes.images, if that's not the case maybe we have an existing issue in the code?

Addison-Stavlo replied:
> if that's not the case maybe we have an existing issue in the code?

Yeah, I'l have to dig a little further to see where that issue is stemming from.  For whatever reason, when we go from 0 images to some images the array passed down in props.attributes.images is redefined, so the uploading images were holding onto an outdated reference (empty array) for the currentMedia of the gallery.  This was causing the bug you noted above when trying to add 2 images to the empty gallery.  It could be something as simple as `images = galleryImages || []` up the chain somewhere, passing us some fallback empty array before things have been defined.  But yeah, if thats the case it will probably be best to clean it up at the source instead.

Addison-Stavlo replied:
I'm not quite sure if there is anything else we can easily do about this one.  It looks like the issue is the default supplied for the image array:

https://github.com/WordPress/gutenberg/blob/0336a3417024537a58cab3114cfbb12a5adabc83/packages/block-library/src/gallery/deprecated.js#L22

So when the gallery is new and has no images, props.attributes.images returns the default value of [].  Once there are images on the gallery, the actual gallery image array is passed instead.  This changing of the array we are referencing is what is causing that issue, the simplest fix I can think of is asserting control over them inside of the gallery component like I did, but am open to other suggestions.  🤔

mkevins replied:
I encountered a similar issue on mobile, with a stale reference to the collection, and used the `useRef` hook to work around it: https://github.com/WordPress/gutenberg/blob/c6a9ab6745c0a97d3b8b70d4c1b5f54d4dd7c979/packages/block-editor/src/components/media-placeholder/index.native.js#L48

Upon further reflection, perhaps it'd be more idiomatic to use a more functional interface for the `value` prop, something like `{ otherProps, getCurrentValue = () => [], } = props;` and then used like: `<MediaPlaceholder value={ () => images }`, which would defer resolution of the collection reference, eliminating staleness arising from the asynchronicity. Wdyt?

mkevins replied:
Note: if we go that route, we would need a minor change in `value` prop (e.g. replacing `value={ this.props.attributes }` with `value={ () => this.props.attributes }`) for the following blocks: Video, Gallery, Image, Audio, and the MediaPlaceholder README would need to be updated to reflect the change.

Addison-Stavlo replied:
Thanks for taking a look @mkevins!  That is a very neat implementation of `useRef`!  I was excited to implement it here and then realized `useRef` can only be used within functional components while both our gallery and mediaPlaceholder are class based currently. 
 Additionally, I am unable to find a similar class centric solution using `createRef` that works for this 🤔 .  Setting `this.images` seems to be the class based solution I can get working, and it works very similarly with an array instead of object.

> perhaps it'd be more idiomatic to use a more functional interface... if we go that route, we would need a minor change in value prop for the following blocks: Video, Gallery, Image, Audio, and the MediaPlaceholder

That is a thought, but it would depend on all of those being functional components.  So we would need to do some refactoring across the board where necessary for this.  Additionally, the only reason I see a need arising here is because the Gallery displays multiple media items at a time which makes this uploading process a bit more complex if we want dynamic behavior throughout that process.  Blocks that only display 1 media item don't really have the need for something like this as far as I can tell?

mkevins replied:
Hi @Addison-Stavlo :wave: I think my suggestion may have been misinterpretted / misunderstood, (I could have worded it better :sweat_smile: )

> That is a thought, but it would depend on all of those being functional components. So we would need to do some refactoring across the board where necessary for this.

To clarify, the `useRef` workaround I had used in the past is the functional-component equivalent of exactly your class-component solution (i.e. the `useRef` is analogous to using `this.images` instance property). What I meant for an "alternative" solution is that instead of passing the array reference as the `value` prop to MediaPlaceholder (which gets stale), we could instead pass a function called `getCurrentValue`, which returns the images array when invoked (not stale). This should be possible for both functional components and class-components.

<details>
<summary>So onFilesUpload would look something like this:</summary>

```jsx
	onFilesUpload( files ) {
		const {
			addToGallery,
			allowedTypes,
			mediaUpload,
			multiple,
			onError,
			onSelect,
			// value = [],
			getCurrentValue = () => [],
		} = this.props;
		let setMedia;
		if ( multiple ) {
			if ( addToGallery ) {
				// const currentValue = value;
				setMedia = ( newMedia ) => {
					// onSelect( currentValue.concat( newMedia ) );
					onSelect( [ ...getCurrentValue(), ...newMedia ] );
				};
			} else {
				setMedia = onSelect;
			}
		} else {
			setMedia = ( [ media ] ) => onSelect( media );
		}
		mediaUpload( {
			allowedTypes,
			filesList: files,
			onFileChange: setMedia,
			onError,
		} );
	}
```
</details>

<details>
<summary>And Gallery's edit.js mediaPlaceholder would look something like this:</summary>

```jsx
		const mediaPlaceholder = (
			<MediaPlaceholder
				addToGallery={ hasImagesWithId }
				isAppender={ hasImages }
				className={ className }
				disableMediaButtons={ hasImages && ! isSelected }
				icon={ ! hasImages && sharedIcon }
				labels={ {
					title: ! hasImages && __( 'Gallery' ),
					instructions: ! hasImages && PLACEHOLDER_TEXT,
				} }
				onSelect={ this.onSelectImages }
				accept="image/*"
				allowedTypes={ ALLOWED_MEDIA_TYPES }
				multiple
				// value={ hasImagesWithId ? images : undefined }
				getCurrentValue={ () => images }
				onError={ this.onUploadError }
				notices={ hasImages ? undefined : noticeUI }
				onFocus={ this.props.onFocus }
			/>
		);
```
</details>

Does that help clarify?

Addison-Stavlo replied:
Oh gotcha, thanks for clarifying!  I do really like how that works.

That works in this case given that `getValue` is defined with this.props.attributes.images explicitly and not the dereferenced version.  `getCurrentValue={ () => this.props.attributes.images }`.  Then in the setMedia function I can call the getValue function instead of using the `currentMedia` I have defined there.

I do like this solution you are proposing, but I wonder if there would be any potential problems converting things to work this way?  Would we need backwards compatibility to support any 3rd party blocks that have been created using the value so they wouldn't break when we change this to be using the getValue?  In which case we should probably keep this current solution as well to provide this desired functionality in that case too?  If thats the case it may not be worth going down that road? 🤔

mkevins replied:
> Would we need backwards compatibility to support any 3rd party blocks that have been created using the value so they wouldn't break when we change this to be using the getValue

Hm.. :thinking: I think we can use `value` as well, if needed, to support backward compatibility with third party blocks that have already been implemented using the current interface. Maybe something like this: `{ otherProp, value = [], getCurrentValue = () => value }` (we can use the `value` parameter in the default value for the `getCurrentValue` parameter, as long as it is listed _after_ the value parameter. IMO, this should satisfy backward compatibility for any previous 3rd party implementations.

Addison-Stavlo replied:
Ah yeah, we could do something like that.  I think in this case it would need to return this.props.value as returning the dereferenced version will become stale, but either way.. that will work!  So we could (fairly easily) refactor things to work this way and support backwards compatibility as well.

That would mean we could limit that change/refactor to just the placeholder and gallery block, and all other blocks could still work as is.  Code wise it does seem like the best solution, but I wonder what others will think about the necessity of this change.  Would adding extra props and updating the docs with this other option be worthwhile considering its only for 1 block that we already have an alternative working solution for that does not require that?  🤷‍♀️ Either way,  I'l branch off this PR and create that solution later today so we can have both available for consideration.  

Thanks again for your help here!

jorgefilipecosta replied:
I think I'm missing something in the problem with arrays changing.

> when we go from 0 images to some images the array passed down in props.attributes.images is redefined

> I encountered a similar issue on mobile, with a stale reference to the collection, and used the useRef hook to work around it


I guess it is expected that a value changes, and from one invocation to another, and a different value (reference) is passed. For example, if a user of the component passes a random array (redefined) of images every second, the component should update and work after each change. Could the issue be that somewhere in the code, we are relying on value always being the same reference? It would mean that the issue would not be that the array is redefined but that some code is relying on having a constant reference.

Addison-Stavlo replied:
> Could the issue be that somewhere in the code, we are relying on value always being the same reference? It would mean that the issue would not be that the array is redefined but that some code is relying on having a constant reference.

@jorgefilipecosta You are correct.  In order to get the desired dynamic functionality I added code that required the value always having the same reference.  It sounds like @mkevins did something similar and got around that by using the `useRef` hook, while here I got around that by using `this.images` since we aren't in a functional component.    

However, the new solution we were discussing now changes that.  Instead of having code that requires the reference to not change, this has been refactored to use a `getValue` function instead which eliminates the need to do all that mess that was going on with `this.images`.  I have just updated this (See e8615b4).

jorgefilipecosta commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -110,24 +110,54 @@ export class MediaPlaceholder extends Component {
 		} = this.props;
 		let setMedia;
 		if ( multiple ) {
-			if ( addToGallery ) {
```
The add to gallery flag indicates that the media upload should be added to the existing media items, while if the flag is false the media uploaded should totally replace the existing media.
It seems we are removing the mechanism to totally replace existing media, should we keep the `if ( addToGallery ) {` condition and move our code inside the if?

Addison-Stavlo replied:
Since we are removing this condition, we are always assuming `addToGallery` is true here.  So we are always adding to the gallery, and never ‘replacing’.  If we look at how `addToGallery` prop is defined in `gallery/edit.js`, it is true whenever there are images in the gallery that contain IDs.  

This is problematic, consider the flow where we start with an empty gallery.  We add image1, `addToGallery` will be `false`.  While image1 is still uploading, we drag and drop image2 to start uploading.  Since image1 is not yet uploaded, there are no images with IDs in the gallery, and image2 starts its upload with `addToGallery` as false as well.  That is even though there is already an image in the gallery uploading (and our expectation for image2 is to ADD it to the gallery as opposed to replacing content), the second upload starts with the assumption that we are replacing content instead of adding.

Moreover, `addToGallery` in this implementation seems unnecessary in that its goal is to determine whether to ‘replace’ or ‘add’ the content, while the only time that it signifies replacing content is when the gallery is empty.  Adding to the gallery makes just as much sense as replacing in this circumstance.

I am a bit uncertain of the last place it is being used where it is passed into the `MediaUpload` component from the `MediaPlaceholder`.  It seems it wasn’t necessary to make any changes here to fix the behavioral issues when adding separate uploads to the empty gallery.  This uncertainty is one reason I removed that condition from the setMedia function instead of altering its definition in the gallery component to always be true (maybe it would make sense to do this instead?).

Addison-Stavlo replied:
>  instead of altering its definition in the gallery component to always be true (maybe it would make sense to do this instead?).

Actually, the more I think about it the more this makes sense.  So I added the check for `addToGallery` back into the `media-placeholder` component, and have edited the value we pass into this prop in `gallery/edit.js` to always be `true` for the reasons noted above.

jorgefilipecosta replied:
This solution seems good 👍

jorgefilipecosta commented on packages/block-editor/src/components/media-placeholder/index.js:
```diff
@@ -110,24 +110,54 @@ export class MediaPlaceholder extends Component {
 			multiple,
 			onError,
 			onSelect,
-			value = [],
```
Previously value was not required and the default was empty. Now we are using a default value and we have the following code `const currentMedia = this.props.value;` and then `const filteredMedia = currentMedia.filter( ( item ) => {` this change will make all the existing plugins using MediaPlaceholder without passing a value (were relying on the previous default value) crash because in that case this.props.value is undefined and then we call undefined(currentMedia).filter.

Addison-Stavlo replied:
Good call.  I re-added the fallback for the empty array.  I had removed it due to changes I made in how the Gallery component passes down the value, but it makes sense to keep it regardless of that.

jorgefilipecosta reviewed (commented):
Hi @Addison-Stavlo, I'm sorry for the delay in this review, I had too many requests/notifications, and it took a while to arrive at this one.

Thank you for all the updates from the functionality point of view, things are working very well on the gallery. 👍 

I think we may have some problems in blocks that were relying on previous defaults, and I also left some questions. Thank you in advance for any insights you may provide, and thank you for all the updates.

Addison-Stavlo committed:
re-added fallback value for currentMedia, fixed linter == problem

Addison-Stavlo committed:
updated comment

Addison-Stavlo committed:
re-added addToGallery check, forced 'true' for our gallery

mkevins reviewed (commented):
Hi @Addison-Stavlo :wave: :smiley: I've tested this on mobile (Android) and haven't encountered any regressions. I tried adding images from the media library as well as uploading from device, and with and without existing images within the gallery. All of those flows are working as expected. I left a suggestion for an alternative way of handling the stale collection reference, since I encountered something similar while developing the mobile implementation.

Addison-Stavlo committed:
refactored to use a getValue instead

Addison-Stavlo committed:
updated comment

jorgefilipecosta committed:
Remove getValue and use a reference to this.props.value

jorgefilipecosta reviewed (commented):
Hi @Addison-Stavlo, I added a PR that removes the need to getValue, the difference compared to what we tried before was to reference this.props.value directly inside setMedia if we use a value variable captured before setMedia is defined the behavior is buggy.
Would you be able to confirm if things continue to work with this change? Thank you in advance :)

Addison-Stavlo commented:
@jorgefilipecosta Thanks for taking a look and proposing a cleaner solution.  This "mostly" works.  There is a small behavioral regression on an edge case but it is still a good improvement to the functionality on master.

Using proposed `this.props.value`:
![using-props](https://user-images.githubusercontent.com/28742426/81341744-0dcd6680-9080-11ea-90d6-9f7ea142e838.gif)
As you may notice once the first image finishes uploading, the placeholder for the second image disappears.  Once completed, both images are present in the gallery so we do achieve the correct final state.

Using `getValue`:
![using-getValue](https://user-images.githubusercontent.com/28742426/81341761-132ab100-9080-11ea-84bf-52b8b53aab36.gif)
The placeholder for the second uploading image never disappears unexpectedly.  I was unsuccessful in getting the behavior to work this way when I tried the other approach.

That being said, this is an edge case that I would not expect many users to encounter, and it does end up with the expected state in the end.  I'd be happy either way, the main reason this PR was opened was to handle the case where someone would try to remove existing images while another upload was in progress (the removed images would reappear once the upload completed).  Beyond that, I have tried to address other edge cases that I have found along the way and have probably gotten a bit distracted trying to solve those as well.

jorgefilipecosta reviewed (approved):
This solution addresses most of the issues and does not change anything on the public API. I guess we can go with this solution 👍 
Thank you for all the work and for debugging on this PR @Addison-Stavlo!

Addison-Stavlo committed:
comment update

Addison-Stavlo commented:
Thanks for all of your time and help on this one @jorgefilipecosta ! 

Tested again, things work well!  I updated a comment that was still referencing `getValue`, but other than that I think this is good to go.  Will merge after Travis approves again. 😄

talldan commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -355,11 +355,10 @@ class GalleryEdit extends Component {
 		const mediaPlaceholder = (
 			<MediaPlaceholder
-				addToGallery={ hasImagesWithId }
+				addToGallery={ true }
```
It looks like this might is the cause of https://github.com/WordPress/gutenberg/issues/22338, that when first adding a gallery block the Media Library is always in `addToGallery` mode with all images selected on the Insert Gallery page.

talldan replied:
I've read more of the details on this PR, and I see that the issue is related to `hasImagesWithId` being false when an image upload is in progress and other operations causing the uploading image to be replaced.

I've made naive PR at https://github.com/WordPress/gutenberg/pull/22659.
